ArduCopter/mode_acro.cpp

```diff
@@ -96,46 +96,30 @@ float ModeAcro::throttle_hover() const
 
 // get_pilot_desired_angle_rates - transform pilot's roll pitch and yaw input into a desired lean angle rates
 // returns desired angle rates in centi-degrees-per-second
-void ModeAcro::get_pilot_desired_angle_rates(int16_t roll_in, int16_t pitch_in, int16_t yaw_in, float &roll_out, float &pitch_out, float &yaw_out)
+void ModeAcro::get_pilot_desired_angle_rates(float roll_in, float pitch_in, float yaw_in, float &roll_out, float &pitch_out, float &yaw_out)
 {
     float rate_limit;
-    Vector3f rate_ef_level, rate_bf_level, rate_bf_request;
+    Vector3f rate_ef_level_cd, rate_bf_level_cd, rate_bf_request_cd;
 
     // apply circular limit to pitch and roll inputs
     float total_in = norm(pitch_in, roll_in);
 
-    if (total_in > ROLL_PITCH_YAW_INPUT_MAX) {
-        float ratio = (float)ROLL_PITCH_YAW_INPUT_MAX / total_in;
+    if (total_in > 1.0) {
+        float ratio = 1.0 / total_in;
         roll_in *= ratio;
         pitch_in *= ratio;
     }
 
-    // range check expo
-    g.acro_rp_expo = constrain_float(g.acro_rp_expo, -0.5f, 1.0f);
-    
     // calculate roll, pitch rate requests
-    if (is_zero(g.acro_rp_expo)) {
-        rate_bf_request.x = roll_in * g.acro_rp_p;
-        rate_bf_request.y = pitch_in * g.acro_rp_p;
-    } else {
-        // expo variables
-        float rp_in, rp_in3, rp_out;
-
-        // roll expo
-        rp_in = float(roll_in)/ROLL_PITCH_YAW_INPUT_MAX;
-        rp_in3 = rp_in*rp_in*rp_in;
-        rp_out = (g.acro_rp_expo * rp_in3) + ((1.0f - g.acro_rp_expo) * rp_in);
-        rate_bf_request.x = ROLL_PITCH_YAW_INPUT_MAX * rp_out * g.acro_rp_p;
-
-        // pitch expo
-        rp_in = float(pitch_in)/ROLL_PITCH_YAW_INPUT_MAX;
-        rp_in3 = rp_in*rp_in*rp_in;
-        rp_out = (g.acro_rp_expo * rp_in3) + ((1.0f - g.acro_rp_expo) * rp_in);
-        rate_bf_request.y = ROLL_PITCH_YAW_INPUT_MAX * rp_out * g.acro_rp_p;
-    }
+    
+    // roll expo
+    rate_bf_request_cd.x = g2.acro_rp_rate * 100.0 * input_expo(roll_in, g.acro_rp_expo);
```

Comment:
no need for any protection against invalid values?

Reply:
Protection is contained in the input_expo function.